File: csharp/src/RespBench/Metrics/IntervalMetricsLogger.cs
/*
 * Copyright 2025 the original author or authors.
 */
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Nodes;
using HdrHistogram;
using HdrHistogram.Encoding;
using HdrHistogram.Utilities;

namespace RespBench.Metrics;

/// <summary>
/// Writes per-interval latency histograms to an NDJSON sidecar file.
///
/// Each command gets its own LongConcurrentHistogram for lock-free concurrent recording.
/// A background timer periodically swaps the active histogram with a fresh one and writes
/// the interval snapshot as a single NDJSON line.
/// </summary>
public class IntervalMetricsLogger
{
    private const long MaxLatencyMicros = 600_000_000L;

    private readonly ConcurrentDictionary<string, HistogramSlot> _slots = new();
    private Timer? _timer;
    private string _outputPath = "";
    private string _phaseId = "";
    private long _baseTimeMs;
    private long _lastFlushMs;

    public void Start(string outputPath, string phaseId, int intervalSeconds)
    {
        _outputPath = outputPath;
        _phaseId = phaseId;
        _baseTimeMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        _lastFlushMs = _baseTimeMs;

        // Ensure parent directory exists
        var dir = Path.GetDirectoryName(outputPath);
        if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

        _timer = new Timer(_ => Flush(), null,
            TimeSpan.FromSeconds(intervalSeconds),
            TimeSpan.FromSeconds(intervalSeconds));
    }

    /// <summary>Record a latency value for a command. Thread-safe.</summary>
    public void RecordValue(string commandName, long latencyMicros, bool success)
    {
        var slot = _slots.GetOrAdd(commandName, _ => new HistogramSlot());
        if (success)
            slot.Histogram.RecordValue(Math.Max(1, Math.Min(latencyMicros, MaxLatencyMicros)));
        else
            Interlocked.Increment(ref slot.Errors);
        Interlocked.Increment(ref slot.Requests);
    }

    public void Stop()
    {
        _timer?.Dispose();
        _timer = null;
        Flush();
    }

    private void Flush()
    {
        long now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        long intervalStartMs = _lastFlushMs;
        _lastFlushMs = now;

        var root = new JsonObject
        {
            ["phase_id"] = _phaseId,
            ["interval"] = new JsonObject
            {
                ["start_timestamp"] = DateTimeOffset.FromUnixTimeMilliseconds(intervalStartMs).ToString("o"),
                ["end_timestamp"] = DateTimeOffset.FromUnixTimeMilliseconds(now).ToString("o"),
                ["elapsed_s"] = Math.Round((now - _baseTimeMs) / 1000.0, 1),
                ["duration_s"] = Math.Round((now - intervalStartMs) / 1000.0, 1)
            }
        };

        var metrics = new JsonObject();
        bool hasData = false;

        foreach (var (cmdName, slot) in _slots)
        {
            var fresh = new LongConcurrentHistogram(1, MaxLatencyMicros, 3);
            var snapshot = Interlocked.Exchange(ref slot.Histogram, fresh);
            long requests = Interlocked.Exchange(ref slot.Requests, 0);
            long errors = Interlocked.Exchange(ref slot.Errors, 0);

            if (requests == 0) continue;
            hasData = true;

            var cmdNode = new JsonObject
            {
                ["requests"] = requests,
                ["errors"] = errors
            };

            var latency = new JsonObject
            {
                ["unit"] = "us",
                ["count"] = snapshot.TotalCount
            };

            if (snapshot.TotalCount > 0)
            {
                latency["summary"] = new JsonObject
                {
                    ["min"] = snapshot.GetValueAtPercentile(0),
                    ["p50"] = snapshot.GetValueAtPercentile(50),
                    ["p95"] = snapshot.GetValueAtPercentile(95),
                    ["p99"] = snapshot.GetValueAtPercentile(99),
                    ["p999"] = snapshot.GetValueAtPercentile(99.9),
                    ["max"] = snapshot.GetMaxValue()
                };
                latency["hdr"] = new JsonObject
                {
                    ["format"] = "hdr",
                    ["sigfig"] = 3,
                    ["payload_b64"] = EncodeHistogram(snapshot)
                };
            }

            cmdNode["latency"] = latency;
            metrics[cmdName] = cmdNode;
        }

        if (!hasData) return;

        root["metrics"] = metrics;

        // Memory snapshot
        var proc = Process.GetCurrentProcess();
        var gcInfo = GC.GetGCMemoryInfo();
        root["memory"] = new JsonObject
        {
            ["gc_heap_bytes"] = GC.GetTotalMemory(false),
            ["working_set_bytes"] = proc.WorkingSet64,
            ["private_bytes"] = proc.PrivateMemorySize64,
            ["gc_heap_size_bytes"] = gcInfo.HeapSizeBytes,
            ["gc_committed_bytes"] = gcInfo.TotalCommittedBytes
        };
        string json = root.ToJsonString(new JsonSerializerOptions { WriteIndented = false });
        try { File.AppendAllText(_outputPath, json + Environment.NewLine); }
        catch { /* best-effort */ }
    }

    private static string EncodeHistogram(HistogramBase histogram)
    {
        try
        {
            int neededCapacity = histogram.GetNeededByteBufferCapacity();
            var buffer = ByteBuffer.Allocate(neededCapacity);
            int bytesWritten = histogram.Encode(buffer, HistogramEncoderV2.Instance);
            buffer.Position = 0;
            byte[] bytes = new byte[bytesWritten];
            for (int i = 0; i < bytesWritten; i++)
                bytes[i] = buffer.Get();
            return Convert.ToBase64String(bytes);
        }
        catch { return ""; }
    }

    private class HistogramSlot
    {
        public LongConcurrentHistogram Histogram = new(1, MaxLatencyMicros, 3);
        public long Requests;
        public long Errors;
    }
}

File: php/src/Metrics/IntervalMetricsLogger.php
<?php

declare(strict_types=1);

namespace RespBench\Metrics;

/**
 * Writes per-interval latency histograms, throughput, and memory to an NDJSON sidecar file.
 *
 * Periodically flushes accumulated metrics and resets counters, producing one NDJSON line
 * per interval. Compatible with scripts/plot_intervals.py for visualization.
 */
class IntervalMetricsLogger
{
    /** @var array<string, IntervalCommandSlot> */
    private array $slots = [];
    private string $outputPath;
    private string $phaseId;
    private float $baseTimeMs;
    private float $lastFlushMs;
    private int $intervalSeconds;

    public function start(string $outputPath, string $phaseId, int $intervalSeconds): void
    {
        $this->outputPath = $outputPath;
        $this->phaseId = $phaseId;
        $this->intervalSeconds = $intervalSeconds;
        $nowMs = microtime(true) * 1000;
        $this->baseTimeMs = $nowMs;
        $this->lastFlushMs = $nowMs;

        $dir = dirname($outputPath);
        if (!is_dir($dir)) {
            mkdir($dir, 0755, true);
        }
    }

    /**
     * Record a latency value for a command.
     */
    public function recordValue(string $commandName, int $latencyMicros, bool $success): void
    {
        if (!isset($this->slots[$commandName])) {
            $this->slots[$commandName] = new IntervalCommandSlot();
        }
        $slot = $this->slots[$commandName];
        $slot->requests++;
        if ($success) {
            $slot->latencies[] = $latencyMicros;
        } else {
            $slot->errors++;
        }
    }

    /**
     * Check if it's time to flush and do so if needed.
     * Call this periodically from the benchmark loop.
     */
    public function maybeFlush(): void
    {
        $nowMs = microtime(true) * 1000;
        if (($nowMs - $this->lastFlushMs) >= ($this->intervalSeconds * 1000)) {
            $this->flush();
        }
    }

    public function stop(): void
    {
        $this->flush();
    }

    private function flush(): void
    {
        $nowMs = microtime(true) * 1000;
        $intervalStartMs = $this->lastFlushMs;
        $this->lastFlushMs = $nowMs;

        $durationS = ($nowMs - $intervalStartMs) / 1000.0;
        $elapsedS = ($nowMs - $this->baseTimeMs) / 1000.0;

        $root = [
            'phase_id' => $this->phaseId,
            'interval' => [
                'start_timestamp' => gmdate('Y-m-d\TH:i:s\Z', (int) ($intervalStartMs / 1000)),
                'end_timestamp' => gmdate('Y-m-d\TH:i:s\Z', (int) ($nowMs / 1000)),
                'elapsed_s' => round($elapsedS, 1),
                'duration_s' => round($durationS, 1),
            ],
        ];

        $metrics = [];
        $hasData = false;

        foreach ($this->slots as $cmdName => $slot) {
            if ($slot->requests === 0) {
                continue;
            }
            $hasData = true;

            $cmdNode = [
                'requests' => $slot->requests,
                'errors' => $slot->errors,
            ];

            $latency = [
                'unit' => 'us',
                'count' => count($slot->latencies),
            ];

            if (count($slot->latencies) > 0) {
                sort($slot->latencies);
                $count = count($slot->latencies);
                $latency['summary'] = [
                    'min' => $slot->latencies[0],
                    'p50' => $slot->latencies[(int) ($count * 0.50) - 1] ?? 0,
                    'p95' => $slot->latencies[(int) ceil($count * 0.95) - 1] ?? 0,
                    'p99' => $slot->latencies[(int) ceil($count * 0.99) - 1] ?? 0,
                    'p999' => $slot->latencies[(int) ceil($count * 0.999) - 1] ?? 0,
                    'max' => $slot->latencies[$count - 1],
                ];
            }

            $cmdNode['latency'] = $latency;
            $metrics[$cmdName] = $cmdNode;

            // Reset slot
            $slot->requests = 0;
            $slot->errors = 0;
            $slot->latencies = [];
        }

        if (!$hasData) {
            return;
        }

        $root['metrics'] = $metrics;

        // Memory snapshot
        $phpHeap = memory_get_usage(false);       // PHP emalloc heap
        $phpAllocated = memory_get_usage(true);   // PHP allocated from OS
        $processRss = self::getProcessRssBytes();

        $root['memory'] = [
            'gc_heap_bytes' => $phpHeap,
            'working_set_bytes' => $processRss ?? $phpAllocated,
            'php_allocated_bytes' => $phpAllocated,
            'peak_heap_bytes' => memory_get_peak_usage(false),
            'peak_working_set_bytes' => memory_get_peak_usage(true),
        ];

        $json = json_encode($root, JSON_THROW_ON_ERROR);
        file_put_contents($this->outputPath, $json . "\n", FILE_APPEND);
    }

    /**
     * Get actual process RSS in bytes.
     * On Linux: reads /proc/self/status VmRSS (current RSS).
     * On macOS: uses getrusage() ru_maxrss (peak RSS — no current RSS API).
     */
    private static function getProcessRssBytes(): ?int
    {
        // Linux: read current RSS from /proc/self/status
        if (PHP_OS_FAMILY === 'Linux' && is_readable('/proc/self/status')) {
            $status = @file_get_contents('/proc/self/status');
            if ($status !== false && preg_match('/VmRSS:\s+(\d+)\s+kB/', $status, $m)) {
                return (int) $m[1] * 1024;
            }
        }

        // macOS fallback: ru_maxrss (peak RSS, in bytes on macOS)
        $usage = getrusage();
        if ($usage === false || !isset($usage['ru_maxrss'])) {
            return null;
        }
        return $usage['ru_maxrss'];
    }
}

class IntervalCommandSlot
{
    public int $requests = 0;
    public int $errors = 0;
    /** @var int[] */
    public array $latencies = [];
}
